### ISSUE TITLE
Is cppwinrt developing yet? 

### ISSUE THREAD

**Issue by Azhang74** [Microsoft/cppwinrt#222]
Is cppwinrt developing yet? I am feeling that cppwinrt does'nt update very long time! this project stop? I am looking forward !!! specially XAML full support!

**Comment by walbourn** [comment]
The current C++/WinRT headers on this website match the Windows 10 Creators Update SDK (15063) which is the current public release. The next major update is expected for the Windows 10 Fall Creators Update SDK.

**Comment by Azhang74** [comment]
Is there any sessions about cppwinrt  on cppcon 2017?

**Comment by kennykerr** [comment]
We aren't planning more C++/WinRT talks this year, but there will likely be a panel discussion and I'll be happy to answers any questions there.

**Comment by Azhang74** [comment]
I want to know what about is current status of cppwinrt. I want to use it in my new project. specially XAML full support!  Is there a roadmap?

**Comment by MikeGitb** [comment]
That would be nice indeed. Also, I believe the compiler is not available yet. 

The danger with pushing projects out into the open early but then not sharing regular progress updates or some kind of roadmap is that people start to wonder how committed you actually are in completing, improving and maintaining it in the future. Clang/C2 is e.g. one of those examples that made me weary of Microsoft's / visual studio's beta products.

I fully trust that you are working hard on cppwinrt, but I think, if you want people to invest in that Technologie and give you feedback, you need to give them the feeling that progress is happening and the project has a future.

**Comment by Azhang74** [comment]
**_> From Kenny kerr_**

TIMELINE: I wanted to provide some general guidance for the development plan for C++/WinRT. While I won't provide exact dates, as they are subject to change, here's the overview roadmap for upcoming releases, ordered from soonest to sometime later.
We will remove the 10.0.14393.0 headers. You can retrieve them, if needed, by syncing to a previous commit.
We will update the 10.0.15063.0 headers to our latest build. This build has full support for both consuming WinRT classes and implementing WinRT classes. One item of particular note is that the headers provide appropriate support for the generic WinRT types (e.g. IVector<MyStruct) obviating the need to determine the magic appropriate GUID value. 
As of this update, you must compile the C++/WinRT header files using the Visual Studio 2017 15.3 (or later) C++ compiler. Prior compiler versions are not sufficiently standards compliant.
Generally, C++/WinRT will stay abreast of C++ standards and require a modern and up-to-date compiler.
For those developers needing to use VS2015, you can always retrieve the previous release by syncing to a previous commit.
The C++/WinRT headers and the cppwinrt.exe compiler itself ... <waiting for the cheers to die down> ... will ship in the Windows SDK. Expect to first see them in a Insider Preview build of the Windows SDK in late summer/early fall.
We are updating the XAML compiler to produce C++/WinRT code behind files.
We are adding C++/WinRT project support to Visual Studio 2017.

when I TIMELINE: I wanted to provide some general guidance for the development plan for C++/WinRT. While I won't provide exact dates, as they are subject to change, here's the overview roadmap for upcoming releases, ordered from soonest to sometime later.
We will remove the 10.0.14393.0 headers. You can retrieve them, if needed, by syncing to a previous commit.
We will update the 10.0.15063.0 headers to our latest build. This build has full support for both consuming WinRT classes and implementing WinRT classes. One item of particular note is that the headers provide appropriate support for the generic WinRT types (e.g. IVector<MyStruct) obviating the need to determine the magic appropriate GUID value. 
As of this update, you must compile the C++/WinRT header files using the Visual Studio 2017 15.3 (or later) C++ compiler. Prior compiler versions are not sufficiently standards compliant.
Generally, C++/WinRT will stay abreast of C++ standards and require a modern and up-to-date compiler.
For those developers needing to use VS2015, you can always retrieve the previous release by syncing to a previous commit.
The C++/WinRT headers and the cppwinrt.exe compiler itself ... <waiting for the cheers to die down> ... will ship in the Windows SDK. Expect to first see them in a Insider Preview build of the Windows SDK in late summer/early fall.
We are updating the XAML compiler to produce C++/WinRT code behind files.
We are adding C++/WinRT project support to Visual Studio 2017.

**# **I am very excited when known this!!!****

**Comment by mustaphakd** [comment]
welcome every bits of it :)

**Comment by MikeGitb** [comment]
That was a fast and comprehensive response ;) thank you very much!
I hope we didn't rush you into anything

**Comment by BrentRe** [comment]
Note a couple of minor corrections. I wrote the timeline, so please don't "blame" Kenny for any errors or omissions. :)

Additionally, Scott and Kenny are now scheduled to talk to CppCon on Friday, September 29 • 4:45pm - 5:45pm talking about C++/WinRT and the future of C++ on Windows.

https://cppcon2017.sched.com/event/d57f3289854bd09b1a972afb37f5b0e2

**Comment by Azhang74** [comment]
@Scott @BrentRe @Kenny kerr thank you very much! I am looking forward to "C++/WinRT and the future of C++ on Windows“.

**Comment by mroz7** [comment]
VS2017.3.3 -> I've just compiled basic XamlCalendar project from provided samples and my feelings are mixed. The only cpp file is App.cpp that has 74 lines of code, so it is hallo word project, and:

1. Compiler creates **930MB** pch 
2. To compile app compiler uses at its peak a **1.2GB RAM**
3. IDE's vcpkgsrv,exe takes almost **500MB** to intellisense this code.

This is only hello world project, I haven't tested nothing about samples, but I'm curious if it's possible to compile at all a quite large project.

And last thought - the worst thing is that programmers are not separated from an Windows' nightmare: COM environment with all its complexity, human unreadable GUIDs, Activators, Registrations, blended with massive "should-be-obsolete" registry entries. All that still exist under the hood and that means no one except MSFT's dev dinosaurs know anything about this ancient and "should-be-obsolete" tech

**Comment by walbourn** [comment]
@mroz7: So the next step is for you to compile a large XAML app and compare it to determine the 'fixed overhead' of the system headers. In general memory on a PC is extremely cheap, disk space is basically free, and in the space vs. time tradeoff for development almost always time wins.

PCH files are quite large, but the time they save is well worth it. The only time I find them onerous is when you have a solution with hundreds of individual projects, each with large PCH files. In that case, one should reorganize the projects.

**Comment by anhoavu** [comment]
@mroz7 Since Microsoft never care about optimization of Visual Studio, we have to live with this for a long run. To add on to @walbourn comment, you can use Visual Studio Team Service and disable PCH completely. (I have to completely disagree with @walbourn  about just "buy a larger build drive": not everyone is floated on money and is using a traditional PC; the people with laptop or Surfaces have limited memory choices.)

To minimize rebuild time and making PCH unnecessary, you can split the code into multiple files and split your projects into testable modules. For instance, have a source file such as `MainPageXaml.cpp` for UI-related stuffs which contains those `#include "winrt/*"` (I don't use XAML files but usually write the code directly for any changes to XAML file will typically initiates rebuild of the whole applications due to dependency) and `MainPage.cpp` for other non-UI logic, use global variables instead of class member and global functions instead of class methods to minimize the amount of changes you need on the header file which initiates re-builds. Even with a quad-core 1.6GHz, 2GB RAM machine, I can develop comfortably with these tricks.

**Comment by mroz7** [comment]
@walbourn - I'm glad you have edited, extended and softened sound of your reply, because I only shared MY first user (programmer) experience about compiling cpp/WinRT basic project and my mixed feelings about that.

930MB pch - You answered that hdd on amazon are cheap but there is not problem with capacity of hdd we use, but problem of heavy I/O needed to generate/use such large pch and memory pressure to compile basic xaml app that makes build time really slow. Kenny Kerr mentioned elsewhere that all cpp/winrt is great stress test for compiler. I think not only for compiler, but for all our PC's hardware. We dont want to buy Azure-like farm servers (or Azure services) to intellisense & compile code, we want work with new API on average PC with which we are able to work comfortably with WinAPI.

The same problem was with large and heavy I/O PDBs and linking with only 4GB process space limit,  and VS team agreed with us and tried addressing this by fastlink switches and now we get one problem somehow resolved and now seems some kind of these problems will hit us again working with cpp/winrt.

Long story short - we really would like to develop on windows comfortably, using fast and easy to use tech without unnecessary complexity. Complexity are obvious in complex solutions, but we still talk about common foundations like UI. And I know that it's to late, you working on tech you have.

Even Larry Ostermann replaying to Kenny Kerr on Twitter that really all COM developers are dinosaurs. I can bet MSFT is only company that extensively produce COM based software in XXI century. In other companies COM-based codebase is like Cobol-codebase a legacy code that have to be somehow maintained with no love.



@anhoavu - thanks, but I cannot recommend in my company such tricks to develop comfortably. We have code guidelines being opposite to some of your tips ;-)

**Comment by MikeGitb** [comment]
@mroz7: Just wondering, have you checked if the example also compiles with clang and how much memor that requires?

**Comment by mroz7** [comment]
@MikeGitb - no, i haven't, only cl. 

BTW, I've read Andrew Pardoe's VS blog where he claimed that VS team have ceased further investing in clang support in VS, so...

**Comment by MikeGitb** [comment]
@mroz7: Yeah, I've read that too and am pretty disappointed about that, but I actually meant the clang/llvm toolchain.

**Comment by anhoavu** [comment]
@MikeGitb @kennykerr I tried to build Syndication sample with this command (clang 5.0.0)

    clang -fms-volatile -fno-builtin -fms-extensions -fms-compatibility -fms-compatibility-version=19.10.25017 -std=c++17 -fcoroutines-ts -isystem %MSVC_INCLUDE%  -isystem %WINSDK_INCLUDE% -I %CPP_WINRT_INCLUDE% -c Syndication/Main.cpp -o Main.obj

but I get the following

    Syndication/Main.cpp:12:28: error: no member named 'await_ready' in
      'winrt::Windows::Foundation::IAsyncOperationWithProgress<winrt::Windows::Web::Syndication::SyndicationFeed,
      winrt::Windows::Web::Syndication::RetrievalProgress>'
    SyndicationFeed feed = co_await client.RetrieveFeedAsync(uri);
                           ^~~~~~~~

Do you have any idea what I should add for it to work? In particular, I don't know what is clang corresponding flag for `/await` mentioned in the documentation.

**Comment by MikeGitb** [comment]
@anhoavu: I haven't actually tried this myself yet. According to https://clang.llvm.org/cxx_status.html, the flag would be `-fcoroutines-ts -stdlib=libc++`.

**Comment by MikeGitb** [comment]
Actually, I don't think clang/llvm comes with libc++ on windows, so this will probably not work :(

**Comment by anhoavu** [comment]
@MikeGitb Thanks for the info. I think there is instructions to get `libc++` on Windows but I don't think it is the main problem. The issue I think is that (if I understand correctly then) coroutine is sort of **compiler dependent** i.e. it totally depends on how the compiler interprets `co_await`, etc. Clang assumes libc++ implementation of `coroutine` while cppwinrt is built on Microsoft's implementation of the feature. Anyway, I attempted to convert the example to use the classical `ppl` and just learnt that it is no longer supported and will not be in the future (according to issue #54 ) ! Thus, gone is our hope for cppwinrt+clang.

    int main()
    {
        init_apartment();
        Uri uri(L"http://kennykerr.ca/feed");
        SyndicationClient client;
        SyndicationFeed feed;
        client.RetrieveFeedAsync(uri); // classical create_task not working here!
        /*for (SyndicationItem item : feed.Items())
        {
            hstring title = item.Title().Text();
	    printf("%ls\n", title.c_str());
        }*/
    }

**Comment by kennykerr** [comment]
Regarding precompiled headers, we are very aware of the challenges introduced by the large precompiled headers. This is something that I have been unsatisfied with for a long time. The precompiled header format is incredibly inefficient and requires a great deal of memory. The solution is C++ modules, which is close to approval as a TS and should be fully supported by Visual C++ hopefully quite soon. We have invested quite heavily in alleviating the problem in the short term with the per-namespace header generation. It is very difficult to do so and is not without its own usability challenges but does help ease the pain until modules arrive in all their glory.

Regarding Clang, we have been able to get C++/WinRT working quite satisfactorily with Clang and will be talking about that next week at CppCon. We'll also publish examples as soon as we get a moment to do so.

**Comment by anhoavu** [comment]
@kennykerr Thanks Kenny for the assurance. Hopefully we will be able to use command line from compilation to packaging `*.appx` and launching app, *san*-VS.

**Comment by mroz7** [comment]
@kennykerr - most important to us is you are aware of problems and have plans to improve building experience. Thanks for reply!

**Comment by nvictor** [comment]
umm, played with winrt again this weekend. glad the size issue is mentioned here. thought i was doing something wrong. built 3 of the samples again just to check (also got Windows Creators update). all was fine but each project sample folder was taking about 1GB each. i haven't paid attention to that until now...

**Comment by smallscript** [comment]
I have figured out how to integrate LLVM Clang 7.0 with Visual Studio 2017 update 15.5.6. v1913 with full support for PDB based debugging using the latest LLVM builds.

i.e., lld-link /DEBUG:GHASH; clang-cl -mllvm -emit-codeview-ghash-section flag to clang-cl.

It is a three step process.

 1. Install latest llvm
 2. Update the toolset.props, toolset.targets in VS to support latest clang
 3. Select the new toolset to use for building your C/C++ or other lang project

As of Visual Studio 2017 update 15.4.5 the Microsoft "experimental" Clang C2 no longer works. Thus, the above fixes are necessary to use clang to compile code that has full PDB (not just CodeView /Z7 limited) debuggability. This also now becomes a more complete suite for portability testing cross-platform builds since you can build and PDB debug for windows using all LLVM components from the clang compiler front end to the LLVM codegen backend and LLVM linker.

Cheers, David

**Comment by raffaeler** [comment]
@smallscript could you provide more detailed instructions?
I guess the generated pdb are in the new "portable" format, right?

**Comment by smallscript** [comment]
Check out January 09, 2018 http://planet.clang.org/
Look at the "Try it out!" section:
> If you're already using **clang-cl** and **lld-link** on Windows today, you can try this out.  There are two flags needed to enable this, one for the compiler and one for the linker:
> To enable the emission of a **.debug$H** section by the compiler, you will need to pass the undocumented `-mllvm -emit-codeview-ghash-section` flag to **clang-cl**  (this flag should go away in the future, once this is considered stable and good enough to be turned on by default).
> To tell **lld-link** to use this information, you will need to pass the `/DEBUG:GHASH` to lld.

You just need to pass the  `-mllvm -emit-codeview-ghash-section` flags in either your _c++ projects_ "Command Line:Additional Options" area, or place them directly in the "toolset.props" file that you created in `C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\VC\VCTargets\Platforms\Win32\PlatformToolsets\LLVM-vs2017` or
`C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\VC\VCTargets\Platforms\x64\PlatformToolsets\LLVM-vs2017`.

The key is that in adding those **cli** options you're telling **clang** to emit debug information that the **lld* (aka **lld-link**) will understand and use to produce fully populated **PDB** files. Not the _limited_ ones it made prior to the Jan 09, 2018 drops of LLVM 7.0.

**Comment by raffaeler** [comment]
Thank  you @smallscript !

**Comment by mshaikhcool** [comment]
@mroz7 tell me which part of COM sucks ? also what's your measuring yardstick for something sucking  ? 
If the answer basically boils down to "LOoK It'S nOt C ABI tHeReFoRe iT mUsT sUck" then know for a fact it's high time for you to learn something, you have missed 30 years of opportunity. 

> the worst thing is that programmers are not separated from an old Windows' nightmare: 

Old windows nightmare ? nightmare for who ? For "Ignorant suck ass only knows C ABI type" developers ? Yes.

> COM environment with all its complexity, human unreadable GUIDs, Activators, Registrations

Personal Incredulity + Strawmans fallacy. complexity is a subjective term. aware of that ? GUID are designed to unreadable. Ever took the time to learn COM Apartments ? I guess no.

> Even Larry Ostermann replaying to Kenny Kerr on Twitter that really all COM developers are dinosaurs. I can bet MSFT is only company that extensively produce COM based software in XXI century. 

Personal Incredulity Fallacy. When did A ABI Tech started to become good or bad based on  a person's subjective feelings ? 

> In other companies COM-based codebase is like Cobol-codebase a legacy code that have to be somehow maintained with no love. 

Ad populum. Just because **_"other companies"_** don't use COM, therefore COM is bad !
